src/sentry/models/event.py

```diff
@@ -320,75 +310,310 @@ def site(self):
+    def ip_address(self):
+        return self.__dict__['ip_address']
+
+    @property
+    def title(self):
+        return self.__dict__['title']
+
+    @property
+    def culprit(self):
+        return self.__dict__['culprit']
+
+    @property
+    def location(self):
+        return self.__dict__['location']
+
+    # ============================================
+    # Snuba implementations of django Fields
+    # ============================================
+    @property
+    def datetime(self):
+        """
+        Reconstruct the datetime of this event from the snuba timestamp
+        """
+        # dateutil seems to use tzlocal() instead of UTC even though the string
+        # ends with '+00:00', so just replace the TZ with UTC because we know
+        # all timestamps from snuba are UTC.
+        return parse_date(self.timestamp).replace(tzinfo=pytz.utc)
+
+    @property
+    def time_spent(self):
+        return None
+
+    @property
+    def id(self):
+        # Because a snuba event will never have a django row id, just return
+        # the hex event_id here. We should be moving to a world where we never
+        # have to reference the row id anyway.
+        return self.event_id
+
+    def next_event(self, environments=[]):
+        from sentry.utils import snuba
+
+        conditions = [['event_id', '!=', self.id]]
+
+        if len(environments) > 0:
+            conditions.append(['environment', 'IN', environments])
+
+        result = snuba.raw_query(
+            start=self.datetime,  # gte current event
+            end=datetime.utcnow(),  # will be clamped to project retention
+            selected_columns=self.selected_columns,
+            conditions=conditions,
+            filter_keys={
+                'project_id': [self.project_id],
+                'issue': [self.group_id],
+            },
+            orderby=['timestamp', 'event_id'],
+            limit=5
+        )
+
+        if 'error' in result or len(result['data']) == 0:
+            return None
+
+        events = [e for e in result['data'] if
+                  e['timestamp'] == self.timestamp and
+                  e['event_id'] > self.event_id or
+                  e['timestamp'] > self.timestamp]
+
+        return SnubaEvent(events[0]) if events else None
+
+    def prev_event(self, environments=None):
+        from sentry.utils import snuba
+
+        conditions = [['event_id', '!=', self.id]]
+
+        if len(environments) > 0:
+            conditions.append(['environment', 'IN', environments])
+
+        result = snuba.raw_query(
+            start=datetime.utcfromtimestamp(0),  # will be clamped to project retention
+            end=self.datetime,  # lte current event
+            selected_columns=self.selected_columns,
+            conditions=conditions,
+            filter_keys={
+                'project_id': [self.project_id],
+                'issue': [self.group_id],
+            },
+            orderby=['-timestamp', '-event_id'],
```

Comment:
This in-memory ordering could be really expensive still, especially if we are selecting all of the `selected_columns`, basically we have to sort every event from now until the end of time in memory and find the earliest one.

I think I may try and extend the logic in https://github.com/getsentry/snuba/blob/master/snuba/split.py to work with these types of queries. TL;DR it just searches a small time range first, so that if we find something in that range we can avoid searching and sorting every other event.